graphql/src/lib.rs

```diff
@@ -0,0 +1,156 @@
+    http::{header::CONTENT_TYPE, Method},
+    web,
+    web::{Payload, Query, ServiceConfig},
+    App, FromRequest, HttpRequest, HttpResponse, HttpServer, Result,
+};
+use coordinator::module::{HandleGraphQlRequest, SessionId};
+use futures::StreamExt;
+pub use handler::handle_gql_query;
+use std::collections::HashMap;
+use std::net::SocketAddr;
+use std::sync::Arc;
+
+// TODO: replace with the real Client.
+pub struct Client {}
+
+impl Client {
+    fn allocate_session(&self, _height: Option<u64>) -> SessionId {
+        123
+    }
+}
+
+pub struct GraphQlRequestHandler {
+    pub handler: Arc<dyn HandleGraphQlRequest>,
+    pub session_needed: bool,
+}
+
+pub struct ServerData {
+    client: Arc<Client>,
+    /// Name to (session_needed, handler)
+    graphql_handlers: HashMap<String, GraphQlRequestHandler>,
+}
+
+impl ServerData {
+    pub fn new(client: Arc<Client>, graphql_handlers: HashMap<String, GraphQlRequestHandler>) -> Self {
+        Self {
+            client,
+            graphql_handlers,
+        }
+    }
+}
+
+async fn handle_request(
+    request: HttpRequest,
+    payload: Payload,
+    path: web::Path<String>,
+    server_data: web::Data<Arc<ServerData>>,
+) -> Result<HttpResponse> {
+    let module_name = path.into_inner();
+    if let Some(GraphQlRequestHandler {
+        session_needed,
+        handler,
+    }) = server_data.graphql_handlers.get(&module_name)
+    {
+        let session = if *session_needed {
+            let height = None;
+            server_data.client.allocate_session(height)
+        } else {
+            0
+        };
+
+        match *request.method() {
+            Method::POST => handle_post(request, payload, &**handler, session).await,
+            Method::GET => handle_get(request, payload, &**handler, session).await,
+            _ => Err(ErrorMethodNotAllowed("GraphQL requests can only be sent with GET or POST")),
+        }
+    } else {
+        Err(ErrorNotFound(format!("Module not found: {}", module_name)))
+    }
+}
+
+const MAX_SIZE: usize = 262_144; // max payload size is 256k
+
+async fn handle_post(
+    request: HttpRequest,
+    mut payload: Payload,
+    handler: &dyn HandleGraphQlRequest,
+    session: SessionId,
+) -> Result<HttpResponse> {
+    let content_type_header = request.headers().get(CONTENT_TYPE).and_then(|hv| hv.to_str().ok());
+    let query: HashMap<String, String> = match content_type_header {
+        Some("application/json") | Some("application/graphql") => {
+            let mut body = web::BytesMut::new();
+            while let Some(chunk) = payload.next().await {
+                let chunk = chunk?;
+                // limit max size of in-memory payload
+                if (body.len() + chunk.len()) > MAX_SIZE {
+                    return Err(ErrorBadRequest("overflow"))
+                }
+                body.extend_from_slice(&chunk);
+            }
+            Ok(serde_json::from_slice(body.as_ref()).unwrap())
```

Comment:
I think this is using a low-level API in the library.
I'll find a high-level API.

Reply:
FYI, I referred to this code https://github.com/actix/examples/blob/master/json/src/main.rs#L31